### PR TITLE
Plane: limit forward throttle during auto transition

### PR DESCRIPTION
**Pull request by kd0aij**
@Hwurzburg I think this would be worth testing. We can make the throttle management more sophisticated if this does what I expect: force throttle to the value spec'd by the TRIM_THROTTLE parameter until the forward transition is complete.

### CODE REVIEW

**Comment by WickedShell** [comment]
Why would we do this? This massively extends transition time for those who don't need it, and doesn't really seem to consider that transition is a very draggy portion of flight and you'd need more then hover throttle to get through it...

**Comment by kd0aij** [comment]
@WickedShell See the warning about excessive current draw on this page:
https://ardupilot.org/plane/docs/quadplane-flying.html
Setting the throttle to cruise here is just to see if this is the right place to manage the power demand.
I guess TECS is calling for max throttle during the forward transition?

**Comment by WickedShell** [comment]
@kd0aij not all quad plane setups have that problem though, and it makes it much worse for those that don't have the problem. (IE gas engines or split battery packs do not have this problem at all), this change makes transitions significantly worse for that class of vehicle. (Again cruise throttle is probably actually not sufficient to get you through a transition to your cruise airspeed due to the extra drag of running the quad motors)

**Comment by kd0aij** [comment]
That's fine. I didn't mean to imply that this WIP PR was anything other than work-in-progress...

**Comment by Hwurzburg** [comment]
@tridge had suggested that it be TKOFF_THR_MAX and that it be ramped to this value over a couple of seconds....perhaps by THR_SLEWRATE/2?

**Comment by WickedShell** [comment]
TKOFF_THR_MAX isn't bad, although the slewing needs to be optional, or have a separate slew rate then the rest of flight.

**Comment by kd0aij** [comment]
I just tried this in SITL with the BigStik quadplane; 75% throttle seems to be enough for it. 
As Michael says, cruise throttle is, of course, too low.

**Comment by kd0aij** [comment]
@Hwurzburg So, anyway, this does appear to be the right place to do the throttle management.

**Comment by Hwurzburg** [comment]
separate slewrate sounds good...the rationale is that having some slew prevents the sag and that by the time the forward motors have some throttle, the vtol motors will have reduced their contributions some....anyway, Mark, its a good start!

**Comment by IamPete1** [comment]
Why not setup BATT_WATT_MAX as it suggests in that warning on the wiki?

**Comment by Hwurzburg** [comment]
because it only moves the throttle lower 10%/sec....battery can be damaged much quicker than that since full throttle is currently applied through transition

**Comment by IamPete1** [comment]
Ah i see, so having a throttle slew rate of less than 10% per sec will allow BATT_WATT_MAX to keep up.

**Comment by Hwurzburg** [comment]
It would if the code were changed to faster slew, but oscillations would probably result....fast back off of throttle, batt volt jumps back up,throttle up, voltage drop, throttle back off, volt up,etc.....the slower slew acts as a bandwidth limiter/LPF on that loop to prevent oscillations...using a throttle constraint instead removes chance for oscillation since it does not have feedback....

**Comment by Hwurzburg** [comment]
@kd0aij Mark, are we sure this is right place?...it would only be for AUTO mode transitions, right? shouldn't it be somewhere in the TRANSITION_AIRSPEED_WAIT update?

**Comment by kd0aij** [comment]
forward throttle in VTOL modes is calculated by QuadPlane::forward_throttle_pct()
In FW modes, I the throttle is calculated  in Plane::calc_throttle() which is called from ModeAuto::update when in auto mode.

**Comment by Hwurzburg** [comment]
duh, been so  on tailsitters recently (they fix throttle on transition) that I forgot what I knew about man throttle in manual transitions on non-tailsitters....sorry...

**Comment by kd0aij** [comment]
@Hwurzburg I just took another pass at this, using TKOFF_THR_MAX and a hardwired 1 second ramp length. I expect 1 second is too short to avoid problems, based on the advice here: https://ardupilot.org/plane/docs/batt-watt-max.html#batt-watt-max
We could throw another new parameter at it, or maybe use Q_TRANSITION_MS/2, since that defaults to 5 seconds.

**Comment by Hwurzburg** [comment]
If THR_SLEWRATE is active, I actually think that just limiting it to TKOFF_THR_MAX would be all that is needed, since the default slewrate would limit it to a second by default...with TKOFF_THR_SLEW also available (although its inactive by default) there is plenty to allow control....I dont think we can idiot proof it.....I think the real issue is that in AUTO now, the transition start slams the forward throttle all the way, instead of adhering to a limit, like TKOFF_THR_MAX...which many do look at and set......I have not had any problems with transition with my 10C 5000mah batts (50A max) with 38A max transitions (75% thr_max)...I think fixing the limit to TKOFF_THR_MAX is good enough..

**Comment by kd0aij** [comment]
Then we could eliminate the ramp that I added, unless you want to guarantee some maximum slew rate, 
@WickedShell What would you prefer?

**Comment by Hwurzburg** [comment]
KISS....I think just using TKOFF_THR_MAX is good enough

**Comment by kd0aij** [comment]
OK, squash the two commits to eliminate the ramp, or delete the second commit to retain it.
Verified in SITL that THR_SLEWRATE is effective... don't know what effect TKOFF_THR_SLEW would have.

**Comment by IamPete1** [comment]
Why not apply this to all forward transitions?

**Comment by Hwurzburg** [comment]
now we have RC control of exactly how much throttle is applied during the transition, which I think is beneficial....in AUTO you dont

**Comment by kd0aij** [comment]
another way of saying that is that modes like FBWA don't assert any control over throttle during a transition; if your throttle stick is low when you switch from a VTOL mode to FBWA, it may take a very long time for the transition to complete.

**Comment by WickedShell** [comment]
I'm sorry I must be missing something here, I don't object to using TKOFF_THR_SLEWRATE and TKOFF_THR_MAX to limit throttle during transitions, but we don't ever want to just stick the throttle to a constant percentage. This was a massive bug and problem in normal fixed wing takeoffs, and I'd hate to see it come back again. This will easily lead to overspeeding the aircraft during the timer wait part of transition.

**Comment by kd0aij** [comment]
Are you saying to call plane.calc_throttle() first, then limit the result?

**Comment by WickedShell** [comment]
TKOFF_THR_MAX actually could be higher then the normal throttle max, that's actually how I always used it. Normally I didn't want to allow the plane that power level but to get through a takeoff/hand launch quickly it was better to have the burst power. If you have a separate battery (or are a gas plane) then the same logic actually may apply to a VTOL transition.

Basically you don't want to bypass TECS at all. Isn't the problem more that you would need to extend the check here: https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/servos.cpp#L415 Worth noting though that the call after this is to actually engage the watt limiter if available which is probably part of what you wanted as a power draw protection in the first place?

**Comment by Hwurzburg** [comment]
The original request from @tridge was to get some kind of limit on throttle in AUTO transitions...since there are none now....slew rate limiting, watt limiting, etc. are all already there and mentioned as possible ameliorating things, but it still goes to MAX in AUTO pretty quickly in default setup...I always tend toward KISS where possible...

**Comment by WickedShell** [comment]
Fixed throttle doesn't work though. It's an objectively terrible solution, that introduces other problems on a wide number of setups. If you look at the relavant linked code though I think it would be quite achievable to hook both that and the other limiting to be using the TKOFF_THR values, which would allow you to set a lower throttle max, and a lower slewrate (which would probably be the main improvement).

I don't disagree that we can improve throttle management, but hard coding the throttle output for the duration of the entire transition is a regression.

**Comment by kd0aij** [comment]
@WickedShell Thanks, I hadn't realized that both the update_control_mode and set_servos tasks call SRV_Channels::set_output_scaled(SRV_Channel::k_throttle,...) at 400Hz.
I wonder why calc_throttle isn't called from set_servos instead?

**Comment by kd0aij** [comment]
Maybe all that is needed is to make calc_throttle respect THR_MAX?
It seems like that would keep TECS from ever calling for 100%, but I don't know whether you would consider that good or bad :)

I changed this PR to retain the previous behavior (calling calc_throttle first) and then enforce the throttle limit(s). It should now only change behavior w.r.t. forward transitions in auto mode iff TECS calls for more than the limit.
Might be worth testing at this point @Hwurzburg

**Comment by kd0aij** [comment]
@Hwurzburg I just tested this in a sitl auto mission with Tridge's Realflight "Big Stik QuadPlane APM -  rev7_AV" and throttle is limited on forward transition to TKOFF_THR_MAX (if set) else to THR_MAX which defaults to 100%.

There should be no effect in any other state, the change only applies when in auto mode and QuadPlane::in_transition returns true.

The only remaining question is whether a new parameter is necessary instead of  using TKOFF_THR_MAX/THR_MAX

**Comment by Hwurzburg** [comment]
I think as is will be fine for a start....its the same limit on forward throttle as if you hand launched in fixed wing mode on auto or takeoff mode and not now full throttle(if setup properly)....of course,ideally, since the current will be both the forward throttle + the VTOL throttle for hover (initially, decreasing rapidly), to limit it to the same max current as plain fixed wing takeoff, it  would need a separate limit to allow for that additional initial VTOL contribution... but something is better than nothing!

**Comment by Hwurzburg** [comment]
Mark, this has been sitting around for a while...want to put a devcall label on it and get it in?

File: ArduPlane/mode_auto.cpp [hunk 1]
```diff
@@ -76,6 +76,16 @@ void ModeAuto::update()
         } else {
             plane.calc_throttle();
         }
+    } else if (plane.quadplane.in_transition()) {
```
Comment (tridge) [inline ArduPlane/mode_auto.cpp:79]: I think we can just treat transition like takeoff in Plane::set_servos_controlled()

Comment (kd0aij) [reply ArduPlane/mode_auto.cpp:79]: Do you mean this should be moved to set_servos_controlled?

**Comment by kd0aij** [comment]
This is just a 1-line change when done in set_servos_controlled as suggested.
Tested in SITL with a quadplane auto mission; throttle limited to THR_MAX on forward transition.

**Pull request merged**